### ISSUE TITLE
Enhance Magic Dot to support iteration

### ISSUE THREAD

**Issue by bonafideduck** [bonafideduck/magic_dot#8]
* Magic Dot version: 0.1.1
* Python version: 3
* Operating System: all

### Description

Enhance Magic Dot to support iteration.   They typical response will return a MagicDot wrapped contained data.  There are some situations to consider.  By default if the wrapped data is:

* Iterable Data: Return an iterable similar to [ MagicDot(d, **settings_from_d) for d in md.__data ]
* NOT_FOUND: Raise a NotFound exception
* Not Iterable: Raise a TypeError

The latter two can be configured with the following:

* iterator_not_found = 
    * "exception" - the default
    * "empty" - an empty list
    * "self" - return a [MagicDot(md.__data__)]

* iterator_type_error = (same three as above)

Several design decisions should be mulled over.
* For iterating over a NOT_FOUND, should we limit it to two things, if exception() is enabled raise the NotFound exception and otherwise, return a "self" like above.  Leaving out the "empty" option.

* Or perhaps we should drop the iterator_not_found handling and if exception is set, raise a NotFound exception otherwise do what the iterator_type_error specifies.  (I'm liking this choice.)

* Should the parameter be an Enum instead of simple strings.  It adds verbosity but forces good behavior.  (I also need to better understand Enums, they always didn't seem what I expected.) Maybe a compromise:  "ite = IteratorTypeError.exception"

**Comment by bonafideduck** [comment]
I think I'll go for a bit simplified.

* If .exception() is set, this causes an exception during the extraction phase but not the iteration.  Note that the below has two list items where the second does not contain a 'a' key.
```
>>> ab = MagicDot([ {'a':1}, {'b':2} ])

>>> ab.a.get()
[1, NOT_FOUND]

>>> ab.exception().a.get()
raises magic_dot.exceptions.NotFound

>>> [ i for i in ab.a ]
[MagicDot(1), MagicDot(NOT_FOUND)]

>>> [ i for i in ab.a.exception() ]
raises magic_dot.exceptions.NotFound
```

* Normally if iteration is attempted over a non-iterable, the defautl of TypeError is returned.  If .id_for_non_iter() is set, then if iteration is attempted for non-iterable data, it returns [MagicDot(data.get())].

```
>>> one = MagicDot(1)

>>> [ i for i in one ]
TypeError: 'int' object is not iterable

>>> [ i for i in one.self_for_non_Iter() ]
Out: [MagicDot(1)]
```

**Comment by bonafideduck** [comment]
Upon further consideration, I think I have settled on a solution.

There are two exception cases:
1. The wrapped data is not iterable.  Attempting to iterate over valid data will always raise a TypeError exception.
2. The wrapped data is NOT_FOUND.  Attempting to iterate will either:
    1. Raise TypeError (this is the default) or
    2. Return an empty iterator `iter(())` if iter_nf_as_empty() is configured.

**Comment by bonafideduck** [comment]
This was released with version 0.2.0.
